File: project_euler/even_fibonacci_numbers.ts
const sumOfEvenNumbers = (limit: number, series: number[] = [], index = 0, sumOfEven = 0): number => {
	if (index < 1) series.push(0) 
	else if (index <= 2) series.push(1) 
	else {
		const nextNumberInSeries = series[series.length - 1] + series[series.length - 2]
		series.push(nextNumberInSeries)
		if (nextNumberInSeries > limit) {
			return sumOfEven
		}
	}
	if (!(series[series.length - 1] % 2)) sumOfEven += series[series.length - 1]
	return sumOfEvenNumbers(limit, series, index + 1, sumOfEven)
}

console.log(sumOfEvenNumbers(8))
console.log(sumOfEvenNumbers(400))
console.log(sumOfEvenNumbers(4000000))

File: project_euler/even_fibonacci_numbers.dart
// https://projecteuler.net/problem=2
// https://www.geeksforgeeks.org/dsa/even-fibonacci-numbers-sum/
void main() {
	print(sumOfEvenNumbers(8, <int>[]));
	print(sumOfEvenNumbers(400, <int>[]));
	print(sumOfEvenNumbers(4000000, <int>[]));
}

int sumOfEvenNumbers(int limit, List<int> series, {int index = 0, int sumOfEven = 0}) {
	if (index == 0) {
		series.add(0);
	} else if ([1, 2].contains(index)) {
		series.add(1);
	} else {
		series.add(series[series.length - 1] + series[series.length - 2]);
		if (series.last > limit) {
			return sumOfEven;
		}
	}
	if (series.last % 2 == 0) {
		sumOfEven += series.last;
	}
	return sumOfEvenNumbers(limit, series, index: index + 1, sumOfEven: sumOfEven);
}